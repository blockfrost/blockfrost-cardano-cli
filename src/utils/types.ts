export type Awaited<T> = T extends PromiseLike<infer U> ? U : T;
export type CommandDataType<T> = T extends (...args: any) => any ? Awaited<ReturnType<T>> : unknown;
