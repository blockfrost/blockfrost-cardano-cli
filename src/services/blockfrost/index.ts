import { BlockFrostAPI } from '@blockfrost/blockfrost-js';
import { ERROR } from '../../constants/errors';
import { format } from 'util';
import { ENV_VAR_PROJECT_ID } from '../../constants';
import { CardanoNetwork } from '@blockfrost/blockfrost-js/lib/types';
const packageJson = require('../../../package.json');

export const createBlockfrostClient = (network?: CardanoNetwork) => {
  const envVarName = network ? ENV_VAR_PROJECT_ID[network] : ENV_VAR_PROJECT_ID.mainnet;
  const projectId = process.env[envVarName];

  if (!projectId) {
    throw new Error(format(ERROR.ENV_PROJECT_ID_NOT_SET, envVarName));
  }

  const userAgent = `${packageJson.name}@${packageJson.version}`;

  return new BlockFrostAPI({
    projectId,
    network,
    userAgent,
  });
};
