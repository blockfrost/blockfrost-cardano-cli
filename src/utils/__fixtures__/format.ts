export const bech32ToHex = [
  {
    description: 'Testnet address',
    stripPrefix: false,
    data: 'addr1qx7ay9achkszevnkfnzk450q3lq70wetn3m34pkr340vyaws4dsc4x2tj40vqypwyl75ehkhf4kegl90gxz47psf7dns58yqpa',
    result:
      '01bdd217b8bda02cb2764cc56ad1e08fc1e7bb2b9c771a86c38d5ec275d0ab618a994b955ec0102e27fd4cded74d6d947caf41855f0609f367',
  },
  {
    description: 'Mainnet address',
    stripPrefix: false,
    data: 'addr1vy740r73x2w3du2xxt76cs4hdml4zw2c5h7tddcyf3jauys9tyns4',
    result: '613d578fd1329d16f14632fdac42b76eff513958a5fcb6b7044c65de12',
  },
  {
    description: 'Testnet address',
    stripPrefix: true,
    data: 'addr_test1qpmtp5t0t5y6cqkaz7rfsyrx7mld77kpvksgkwm0p7en7qum7a589n30e80tclzrrnj8qr4qvzj6al0vpgtnmrkkksnqd8upj0',
    result:
      '76b0d16f5d09ac02dd1786981066f6fedf7ac165a08b3b6f0fb33f039bf76872ce2fc9debc7c431ce4700ea060a5aefdec0a173d8ed6b426',
  },
  {
    description: 'Testnet stake address',
    stripPrefix: true,
    data: 'stake_test1urw3nu0jaxhlnw5heuhp48cgqfwpu9tdup6ay2kfnfmdjcc98fh89',
    result: 'dd19f1f2e9aff9ba97cf2e1a9f08025c1e156de075d22ac99a76d963',
  },
  {
    description: 'mainnet stake address',
    stripPrefix: true,
    data: 'stake1u89hxtuxvfdqda90w2aw2mluxcsgyctfe2lz52n986lrc2cumssr9',
    result: 'cb732f86625a06f4af72bae56ffc3620826169cabe2a2a653ebe3c2b',
  },
  {
    description: 'Testnet pool',
    stripPrefix: false,
    data: 'pool1weu4vlg9t8knma7t2j5y3w2k3vzdr9mtnynd2jhfalwn76nwh48',
    result: '7679567d0559ed3df7cb54a848b9568b04d1976b9926d54ae9efdd3f',
  },
];

export const stripQuotes = [
  {
    description: 'Strip quotes around string',
    data: '"abcde"',
    result: 'abcde',
  },
  {
    description: 'Strip quotes around multiline string',
    data: '"abcd "e" \nfg"hi"',
    result: 'abcd "e" \nfg"hi',
  },
  {
    description: 'Strip quotes on non-string value',
    data: { a: 2 },
    result: { a: 2 },
  },
];

export const getNetworkFromRewardAccount = [
  {
    description: 'getNetworkFromRewardAccount Mainnet account',
    data: 'stake1uyqgnh98k99vknl7yl60le3dtcj0ha5vdjmnegr2ha9aqggcjvclw',
    result: 'Mainnet',
  },
  {
    description: 'getNetworkFromRewardAccount Testnet account',
    data: 'stake_test1uzwty2smxft542kejklelercsdxyze4yw9kwucvmt4qldjqh6d9sf',
    result: 'Testnet',
  },
];

export const stringify = [
  {
    description: '(safe stable) stringify',
    data: {
      z: [1, 2, 3],
      a: 'a',
      b: 1,
      c: {
        bc: BigInt('90071992547409910000'),
        ab: 2,
      },
    },
    result: `{
    \"z\": [
        1,
        2,
        3
    ],
    \"a\": \"a\",
    \"b\": 1,
    \"c\": {
        \"bc\": 90071992547409910000,
        \"ab\": 2
    }
}`,
  },
];

export const transformPoolRelays = [
  {
    description: 'single host address',
    data: [
      {
        ipv4: '64.227.109.12',
        ipv6: null,
        dns: null,
        dns_srv: null,
        port: 3000,
      },
    ],
    result: [
      {
        'single host address': {
          IPv4: '64.227.109.12',
          IPv6: null,
          port: 3000,
        },
      },
    ],
  },
  {
    description: 'single host name',
    data: [
      {
        ipv4: null,
        ipv6: null,
        dns: 'relays.testnet.stakenuts.com',
        dns_srv: null,
        port: 3001,
      },
    ],
    result: [
      {
        'single host name': {
          dnsName: 'relays.testnet.stakenuts.com',
          port: 3001,
        },
      },
    ],
  },
];

export const transformPoolUpdateCert = [
  {
    description: 'transformPoolUpdateCert',
    data: {
      poolHex: 'a5a3ce765f5162548181a44d1ff8c8f8c50018cca59acc0b70a85a41',
      cert: {
        cert_index: 0,
        pool_id: 'pool15k3uuajl2939fqvp53x3l7xglrzsqxxv5kdvczms4pdyzhxm6jf',
        vrf_key: '070da3313205e114984ee06f8a1ca99fe1bb72d4518e0d04d1815bb1c8bbd4c4',
        pledge: '900000000000',
        margin_cost: 0.03,
        fixed_cost: '340000000',
        reward_account: 'stake_test1uzwty2smxft542kejklelercsdxyze4yw9kwucvmt4qldjqh6d9sf',
        owners: ['stake_test1uzwty2smxft542kejklelercsdxyze4yw9kwucvmt4qldjqh6d9sf'],
        metadata: {
          url: 'http://23.234.197.69/mks-metadata-001.json',
          hash: 'd98a03b8aa962d80511d62566df2af415afd9bd03d53cbb0ad457a53d3491f74',
          ticker: 'MKS',
          name: 'MKS Stake Pool',
          description: 'testnet stake pool',
          homepage: 'http://23.234.197.69',
        },
        relays: [
          {
            ipv4: '23.234.197.69',
            ipv6: null,
            dns: null,
            dns_srv: null,
            port: 5001,
          },
        ],
        active_epoch: 173,
      },
    },
    result: {
      cost: '340000000',
      margin: 0.03,
      metadata: {
        hash: 'd98a03b8aa962d80511d62566df2af415afd9bd03d53cbb0ad457a53d3491f74',
        url: 'http://23.234.197.69/mks-metadata-001.json',
      },
      owners: ['9cb22a1b32574aaad995bf9fe478834c4166a4716cee619b5d41f6c8'],
      pledge: BigInt(900000000000),
      publicKey: 'a5a3ce765f5162548181a44d1ff8c8f8c50018cca59acc0b70a85a41',
      relays: [{ 'single host address': { IPv4: '23.234.197.69', IPv6: null, port: 5001 } }],
      rewardAccount: {
        credential: { 'key hash': '9cb22a1b32574aaad995bf9fe478834c4166a4716cee619b5d41f6c8' },
        network: 'Testnet',
      },
      vrf: '070da3313205e114984ee06f8a1ca99fe1bb72d4518e0d04d1815bb1c8bbd4c4',
    },
  },
];
