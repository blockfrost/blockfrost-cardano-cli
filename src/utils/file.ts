import { PathLike, writeFileSync } from 'fs';
import { ERROR } from '../constants/errors';

export const writeToFile = (
  filePath: PathLike | number,
  data: string | NodeJS.ArrayBufferView,
): void => {
  try {
    writeFileSync(filePath, data);
  } catch (err) {
    throw Error(`${ERROR.FILE_WRITE_FAIL}: ${filePath}`);
  }
};
