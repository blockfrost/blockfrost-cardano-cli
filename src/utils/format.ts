import sStringify from 'safe-stable-stringify';

export const stringToBigInt = (value: string | null) => (value ? BigInt(value) : value);

export const stringify = (obj: any) => sStringify(obj, undefined, 4); // safe-stable-stringify properly stringifies BigInt

export const stripQuotes = (str: string) => {
  if (typeof str === 'string' && str[0] === '"' && str[str.length - 1] === '"') {
    return str.slice(1, -1);
  }

  return str;
};
