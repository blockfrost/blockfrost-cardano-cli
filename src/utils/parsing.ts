import { stringToBigInt } from './format';

export const parseAsset = (
  hex: string,
): {
  policyId: string;
  assetName: string;
} => {
  const policyIdSize = 56;
  const policyId = hex.slice(0, policyIdSize);
  const assetNameInHex = hex.slice(policyIdSize);

  return {
    policyId,
    assetName: assetNameInHex,
  };
};

export const isJsonString = (str: string) => {
  try {
    JSON.parse(str);
  } catch {
    return false;
  }

  return true;
};

export const assetsToPolicies = (assets: { quantity: string; unit: string }[]) => {
  // Example output
  //   {
  //     "85be0612bc25a9a238b03b32f36b0c679d2d4cea63ff284989e20c4aab6d5933#0": {
  //         "address": "addr_test1qpk3nptn9v6wv5eec4q3tdmjq074svs6jdp5dhglnrtm9lcg9ltteln2y73vzf9knvpn2uacu7v6wxqlgxg84937kq7qwvj3f5",
  //         "value": {
  //             "dca984a3ab2c7b9691de0cf0e09227e19c06b2899500ed8e9279ab32": {
  //                 "": 500
  //             },
  //             "lovelace": 2700000
  //         }
  //     }
  // }
  const policies: { [k: string]: any } = {};
  for (const asset of assets) {
    if (asset.unit === 'lovelace') {
      policies.lovelace = stringToBigInt(asset.quantity);
    } else {
      const { policyId, assetName } = parseAsset(asset.unit);
      if (!policies[policyId]) {
        policies[policyId] = {};
      }

      policies[policyId][assetName] = stringToBigInt(asset.quantity);
    }
  }

  return policies;
};

export const parseCardanoCliVersion = (versionOutput: string) => {
  const versionRe = new RegExp(/cardano-cli (\d+)\.(\d+)\.(\d+)/);
  const version = (versionOutput.match(versionRe) ?? [])
    .slice(1)
    .map((t: string) => Number.parseInt(t, 10));
  return version;
};
