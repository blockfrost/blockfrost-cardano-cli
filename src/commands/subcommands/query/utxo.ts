/* eslint-disable camelcase */
import { Responses } from '@blockfrost/blockfrost-js';
import { cli } from 'cli-ux';
import { ERROR } from '../../../constants/errors';
import { assetsToPolicies, parseAsset } from '../../../utils/parsing';
import { Subcommand } from '../subcommand';

// cardano-cli response
// TxHash                                 TxIx        Amount
// --------------------------------------------------------------------------------------
// 5d78b544e73b691dd4d6a35fc951b05d6f411499a94362ba85851987d6911976     1        2732158602 lovelace + TxOutDatumNone

// TxHash                                 TxIx        Amount
// --------------------------------------------------------------------------------------
// 6f1b5b3a27cca7ab18eb2c1331520c8a39115bb9e952c7ab3b5a969db6aeb2b4     0        1500000 lovelace + 20 4ef8525f365560af6b685169cac4c86acdd122d683d3400f6ca23467.0504030201 + TxOutDatumNone
// ae03bfd7b583392884b6b5a76481d7e5e9ea7c232d0d27475c58fb4c8c3e6146     1        345047752 lovelace + 5 2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15.48616c6c6f + 30 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518 + 39000 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.414441 + 6370 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.4154414441636f696e + 5000 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.4861707079414441 + 18 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.5465737431 + 20 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.5465737432 + 10 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.5465737433 + 1000 4ef8525f365560af6b685169cac4c86acdd122d683d3400f6ca23467.0001020304 + 980 4ef8525f365560af6b685169cac4c86acdd122d683d3400f6ca23467.0504030201 + 27 9c3f26285634740a52c31e5fcd0686fe51a0acb2dbcd9e565d96b736 + 15 ecd07b4ef62f37a68d145de8efd60c53d288dd5ffc641215120cc3db + TxOutDatumNone

// --out-file
// {
//   "17990929a7f143d9ab7a7390522e7ba6689d3c5770e77b3eebc9c3098c6d9dab#0": {
//       "address": "addr_test1qpjputy7h5crx67ddqq0mctfq65anqr8c5slkv4h38lkkd8r4mmzqnk963v2ctk5es6dt34tqk5l6lxchkup0mnfzdesglvywz",
//       "value": {
//           "1cdf9d5eefbc56128137129afc692145783402a3def7bb27da4028a7": {
//               "434f494e": 5000000
//           },
//           "lovelace": 4000000
//       }
//   },
//   "fd38ad68a6216125ffc0ecb46a49c9d4f95d105218a1ec5de9d57df7fe23141d#0": {
//       "address": "addr_test1qpjputy7h5crx67ddqq0mctfq65anqr8c5slkv4h38lkkd8r4mmzqnk963v2ctk5es6dt34tqk5l6lxchkup0mnfzdesglvywz",
//       "value": {
//           "1cdf9d5eefbc56128137129afc692145783402a3def7bb27da4028a7": {
//               "434f494e": 10000000
//           },
//           "lovelace": 2000000
//       }
//   },
//   "de5132133cdb2d3db05382cecd9df84a6e428760209f703c89e561850c095c00#0": {
//       "address": "addr_test1qpjputy7h5crx67ddqq0mctfq65anqr8c5slkv4h38lkkd8r4mmzqnk963v2ctk5es6dt34tqk5l6lxchkup0mnfzdesglvywz",
//       "value": {
//           "1cdf9d5eefbc56128137129afc692145783402a3def7bb27da4028a7": {
//               "434f494e": 5000000
//           },
//           "lovelace": 3000000
//       }
//   }
// }

// blockfrost-cli response
// TxHash                                                           TxIx   Amount
// ──────────────────────────────────────────────────────────────── ────── ────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────
// 6f1b5b3a27cca7ab18eb2c1331520c8a39115bb9e952c7ab3b5a969db6aeb2b4 0      1500000 lovelace + 20 4ef8525f365560af6b685169cac4c86acdd122d683d3400f6ca23467.0504030201
// ae03bfd7b583392884b6b5a76481d7e5e9ea7c232d0d27475c58fb4c8c3e6146 1      345047752 lovelace + 5 2e2f143b3ccbe339145183dc2a799a469e92ab56e0d5b0bd04f54f15.48616c6c6f + 30 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518 + 39000 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.414441 + 6370 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.4154414441636f696e + 5000 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.4861707079414441 + 18 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.5465737431 + 20 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.5465737432 + 10 34250edd1e9836f5378702fbf9416b709bc140e04f668cc355208518.5465737433 + 1000 4ef8525f365560af6b685169cac4c86acdd122d683d3400f6ca23467.0001020304 + 980 4ef8525f365560af6b685169cac4c86acdd122d683d3400f6ca23467.0504030201 + 27 9c3f26285634740a52c31e5fcd0686fe51a0acb2dbcd9e565d96b736 + 15 ecd07b4ef62f37a68d145de8efd60c53d288dd5ffc641215120cc3db


type Data = {address: string; utxo: Responses['address_utxo_content']}
export class Utxo extends Subcommand {
  prettyPrint = (res: Data) => {
    this.log();
    cli.table(
      res.utxo,
      {
        tx_hash: {
          header: 'TxHash',
          minWidth: 30,
        },
        tx_index: {
          header: 'TxIx',
          minWidth: 7,
        },
        amount: {
          header: 'Amount',
          get: row =>
            row.amount
              .map(a => {
                const { policyId, assetName } = parseAsset(a.unit);
                return `${a.quantity} ${policyId}${assetName ? `.${assetName}` : ''}`;
              })
              .join(' + '),
        },
      },
      { printLine: this.log, 'no-truncate': true },
    );
    this.log();
  };

  toFile(res: Data): void {
    const dumpObj: {
      [k: string]: { address: string; value: ReturnType<typeof assetsToPolicies> };
    } = {};
    res.utxo.forEach(u => {
      dumpObj[u.tx_hash] = {
        address: res.address,
        value: assetsToPolicies(u.amount),
      };
    });
    super.toFile(dumpObj);
  };

  doWork = async (): Promise<Data> => {
    if (!this.options.address) {
      throw new Error(ERROR.FLAG_MISSING_ADDRESS);
    }
    const address = this.options.address;
    const utxo = await this.client.addressesUtxosAll(address, { order: 'desc' });

    return {
      utxo,
      address,
    };
  };
}
