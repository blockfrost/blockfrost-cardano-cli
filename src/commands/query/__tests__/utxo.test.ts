/* eslint-disable camelcase */
import * as fs from 'fs';
import { stdout } from 'stdout-stderr';
import * as blockfrostService from '../../../services/blockfrost';
import { stringify } from '../../../utils/format';
import { Utxo } from '../utxo';

const address =
  'addr_test1qrdx08c0mexd5y0pn3ug0j9k2z2myneqcl7xq7yap8hzl5hdc55q34v2ggxw9hea4rr3rry933a2zdh60v43h237s8ksdhw5yh';
const utxos = [
  {
    tx_hash: '284939d755a0e4545120a16b8b7b819c4623c1ee9d37c188ee0d8f3ad352e251',
    tx_index: 0,
    output_index: 0,
    amount: [
      {
        unit: 'lovelace',
        quantity: '1000000',
      },
      {
        unit: '1cdf9d5eefbc56128137129afc692145783402a3def7bb27da4028a7434f494e',
        quantity: '2',
      },
      {
        unit: '1cdf9d5eefbc56128137129afc692145783402a3def7bb27da4028a7434f494f',
        quantity: '3',
      },
    ],
    block: '15580d878bca7a8d13254dd1d056303299f007f9a0b74a42418b7a73d29ca3a3',
    data_hash: null,
  },
];

let mockedAdressesUtxosAll: any;
describe('query utxo', () => {
  beforeEach(() => {
    mockedAdressesUtxosAll = jest.fn((_address: string) => {
      return utxos;
    });

    jest
      .spyOn(blockfrostService, 'createBlockfrostClient')
      // @ts-ignore partial mock
      .mockImplementation((_testnet?: boolean) => {
        return {
          addressesUtxosAll: mockedAdressesUtxosAll,
        };
      });
  });

  afterEach(() => {
    jest.resetAllMocks();
  });

  it('should print raw response', async () => {
    stdout.start();
    await Utxo.run(['--address', address, '--json']);
    stdout.stop();

    expect(mockedAdressesUtxosAll).toHaveBeenCalledTimes(1);

    const output = stdout.output;
    expect(JSON.parse(output)).toMatchObject({
      utxo: utxos,
      address: address,
    });
    jest.resetAllMocks();
  });

  it('should print pretty response', async () => {
    stdout.start();
    await Utxo.run(['--address', address]);
    stdout.stop();

    expect(mockedAdressesUtxosAll).toHaveBeenCalledTimes(1);

    const output = stdout.output;
    expect(output).toMatchSnapshot();
    jest.resetAllMocks();
  });

  it('should save the response to --out-file', async () => {
    jest.spyOn(fs, 'writeFileSync').mockImplementation((_path, _data) => {
      // do nothing
    });
    await Utxo.run(['--address', address, '--out-file', 'filename']);
    expect(fs.writeFileSync).toHaveBeenCalledTimes(1);

    // could be replaced with expect(fs.writeFileSync).toMatchSnapshot() if we find it too verbose
    expect(fs.writeFileSync).toHaveBeenLastCalledWith(
      'filename',
      stringify({
        '284939d755a0e4545120a16b8b7b819c4623c1ee9d37c188ee0d8f3ad352e251': {
          address:
            'addr_test1qrdx08c0mexd5y0pn3ug0j9k2z2myneqcl7xq7yap8hzl5hdc55q34v2ggxw9hea4rr3rry933a2zdh60v43h237s8ksdhw5yh',
          value: {
            lovelace: 1000000,
            '1cdf9d5eefbc56128137129afc692145783402a3def7bb27da4028a7': {
              '434f494e': 2,
              '434f494f': 3,
            },
          },
        },
      }),
    );
  });
});
