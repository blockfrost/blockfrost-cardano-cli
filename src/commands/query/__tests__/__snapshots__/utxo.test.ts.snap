// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query utxo should print pretty response 1`] = `
"
 TxHash                                                           TxIx   Amount                                                                                                                                                       
 ──────────────────────────────────────────────────────────────── ────── ──────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────────── 
 284939d755a0e4545120a16b8b7b819c4623c1ee9d37c188ee0d8f3ad352e251 0      1000000 lovelace + 2 1cdf9d5eefbc56128137129afc692145783402a3def7bb27da4028a7.434f494e + 3 1cdf9d5eefbc56128137129afc692145783402a3def7bb27da4028a7.434f494f 

"
`;
